### PR TITLE
AArch64 CRefine (without fastpath)

### PR DESCRIPTION
**Pull request by lsf37**
This includes the initial sorrying run and proof completion up to the end of CRefine, but for now without the fast path, which is still in progress.

The fast path files contain 4 sorries which we expect the theory linter to pick up, and they are also the reason the session is still set to `quick_and_dirty`.

This is still a relatively raw state of the proofs with many added FIXMEs, to be done in a separate PR. More substantial review feedback should probably be added as FIXME as well, so that we have a basis to build on first.

### CODE REVIEW

**Comment by lsf37** [comment]
Looks like I messed up something in Wellformed_C the rebase. Will have a look.

File: proof/refine/AARCH64/Arch_R.thy
Comment (corlewis) [inline proof/refine/AARCH64/Arch_R.thy:586]:
```suggestion
  "\<lbrakk> s \<turnstile> cap.ArchObjectCap (arch_cap.FrameCap p R sz d opt);
     pspace_in_kernel_window s; valid_uses s; pspace_aligned s \<rbrakk>
   \<Longrightarrow> pptrBase \<le> p \<and> p < pptrTop"
```

File: proof/crefine/AARCH64/Finalise_C.thy
Comment (corlewis) [inline proof/crefine/AARCH64/Finalise_C.thy:1407]: Are these being schematic_goals a way to avoid magic numbers? If so, is there any reason why `of_nat_ptTranslationBits_NormalPT_T_val` isn't one as well?

File: proof/refine/AARCH64/orphanage/Orphanage.thy
Comment (corlewis) [inline proof/refine/AARCH64/orphanage/Orphanage.thy:2163]: It doesn't really matter, but I'm surprised this wasn't still crunchable.

**Review by corlewis — APPROVED** [review]
Very nice work! Thank you for all of the helpful comments throughout the commit descriptions and proofs.

I've gone through the commit messages, with suggestions below, and done some random sampling of the content. All of my comments are pretty minor.

[8f61618](https://github.com/seL4/l4v/pull/728/commits/8f61618c8dc91e77338f8ebfedee982e1cf73282): aarch64 crefine: Machine_C machine ops from VSpace_C
  - I'm not sure, but should the 'from' be 'for'?

[ab8568a](https://github.com/seL4/l4v/pull/728/commits/ab8568a4bbda9d6913e9f0969faa93fa29218c21): aarch64 crefine: review after sorrying VSpace_C
  - The list of items mixes up tenses

[de60870](https://github.com/seL4/l4v/pull/728/commits/de60870586de908d7b2c3e735b6d1014fa434fc6): aarch64: update IpcCancel_C from RISCV64
  - 's/mak_canonical/make_canonical/'

[a490407](https://github.com/seL4/l4v/pull/728/commits/a490407820feba707c947159f6d44b8558a162c8): aarch64 crefine: state rel.: update IRQ node array size
  - it's not really wrong, but the '.:' in the commit title feels wrong

[84247a6](https://github.com/seL4/l4v/pull/728/commits/84247a6904fd8a7e6088cbc4535fd802c59380ff): aarch64 crefine: consolidate machine op assumptions
  - 's/machin ops/machine ops/'

**Comment by corlewis** [comment]
For future consideration, would these sorts of branches and PRs be easier to work with if they had a single commit at the start that just copied everything in from the source directory? That would make it possible to see what is actually new work.

**Comment by Xaphiosis** [comment]
> For future consideration, would these sorts of branches and PRs be easier to work with if they had a single commit at the start that just copied everything in from the source directory? That would make it possible to see what is actually new work.

I did that in some cases when I needed it for internal review: a copy commit and then an update commit... but for a larger review, a commit wouldn't be enough I think. You'd need a PR that copies stuff, and then the PR to review to go on top of that... I'm open minded to such ideas (and fearful of how, if ever, we can manage MCS reviews), but am not clear on how we'd deploy it. It is indeed verging on unreviewable when you don't know what got copied and what introduced.

**Comment by corlewis** [comment]
> > For future consideration, would these sorts of branches and PRs be easier to work with if they had a single commit at the start that just copied everything in from the source directory? That would make it possible to see what is actually new work.
> 
> I did that in some cases when I needed it for internal review: a copy commit and then an update commit... but for a larger review, a commit wouldn't be enough I think. You'd need a PR that copies stuff, and then the PR to review to go on top of that... I'm open minded to such ideas (and fearful of how, if ever, we can manage MCS reviews), but am not clear on how we'd deploy it. It is indeed verging on unreviewable when you don't know what got copied and what introduced.

We could have two branches, one with just the copy commit and the other with the updates. The PR would then be from the second into the first, with it all being fixed up at the end before being merged.

**Comment by Xaphiosis** [comment]
> > > For future consideration, would these sorts of branches and PRs be easier to work with if they had a single commit at the start that just copied everything in from the source directory? That would make it possible to see what is actually new work.
> > 
> > 
> > I did that in some cases when I needed it for internal review: a copy commit and then an update commit... but for a larger review, a commit wouldn't be enough I think. You'd need a PR that copies stuff, and then the PR to review to go on top of that... I'm open minded to such ideas (and fearful of how, if ever, we can manage MCS reviews), but am not clear on how we'd deploy it. It is indeed verging on unreviewable when you don't know what got copied and what introduced.
> 
> We could have two branches, one with just the copy commit and the other with the updates. The PR would then be from the second into the first, with it all being fixed up at the end before being merged.

Right, that's what I meant by two PRs ... two branches, one PR, then that can be merged and another PR created + rubber-stamped, or the existing PR be re-targeted to a combined branch and that change rubber-stamped. Will need discipline. I'll try remember for the next time we have a PR that introduces new files (e.g. I can try it on fastpath stuff).

**Comment by corlewis** [comment]
> Right, that's what I meant by two PRs ... two branches, one PR, then that can be merged and another PR created + rubber-stamped, or the existing PR be re-targeted to a combined branch and that change rubber-stamped. Will need discipline. I'll try remember for the next time we have a PR that introduces new files that are derived from old files (e.g. I can try it on fastpath stuff).

It's technicalities now but I don't see why there's a second PR or any rubber-stamping needed. The second branch would already have the copy commit in it, so if you retarget to master before merging then I think it would all work out.

**Comment by Xaphiosis** [comment]
> It's technicalities now but I don't see why there's a second PR or any rubber-stamping needed. The second branch would already have the copy commit in it, so if you retarget to master before merging then I think it would all work out.

Good idea. I think we're in agreement in general. What I mean by "rubber-stamping" in that case is changing the target of a PR and then merging it without further review, i.e. a process where you change the product, make assumptions, and then approve without much examination, you know, "she'll be right mate" :)

**Comment by corlewis** [comment]
> Good idea. I think we're in agreement in general. What I mean by "rubber-stamping" in that case is changing the target of a PR and then merging it without further review, i.e. a process where you change the product, make assumptions, and then approve without much examination, you know, "she'll be right mate" :)

Right, then yes, sounds like we're loudly agreeing with each other.

File: proof/crefine/AARCH64/Machine_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Machine_C.thy:177]: yes, it really should be. Thanks for catching that!

File: proof/refine/AARCH64/orphanage/Orphanage.thy
Comment (lsf37) [reply proof/refine/AARCH64/orphanage/Orphanage.thy:100]: It doesn't, it just gets rid of the warning `auto` produces from using `intro: pred_tcb_at'`, because `pred_tcb_at'` is already declared as some kind of intro/dest/elim rule.

File: proof/crefine/AARCH64/Machine_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Machine_C.thy:177]: Ok, this is slightly less bad than I was thinking. It's just a redundant clause, not masking a missing one. The assumption comes from the lemma with the same name on `ARM` (it's a lemma there, because the machine interface goes one deeper), and that lemma has the same redundant clause, which just disappears if you use the lemmas as `[simplified]` for instance. Since that one is a lemma, we can be sure that there is nothing missing or inconsistent, the only possibility is that it's redundant.

Will remove.

File: proof/crefine/AARCH64/Machine_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Machine_C.thy:46]: It should be possible to use `irqInvalid` here, will have a look.

File: proof/refine/AARCH64/ArchAcc_R.thy
Comment (lsf37) [reply proof/refine/AARCH64/ArchAcc_R.thy:699]: just a left-over, will fix.

File: proof/crefine/AARCH64/Finalise_C.thy
Comment (lsf37) [reply proof/crefine/AARCH64/Finalise_C.thy:1407]: Most likely reason is that I missed it when I added the `numeral` guards :-)

File: proof/crefine/AARCH64/Machine_C.thy
Comment (corlewis) [reply proof/crefine/AARCH64/Machine_C.thy]: Oh of course, I should have noticed that it was the same as the first assumption. It looks like similar lemmas are inconsistent with whether the equivalent of this is at the start or end of their assumptions, and my guess is that on ARM this was a copy-paste error which combined both options.

Comment (lsf37) [reply proof/crefine/AARCH64/Machine_C.thy]: Yes, that's probably what happened.

File: proof/refine/AARCH64/orphanage/Orphanage.thy
Comment (lsf37) [reply proof/refine/AARCH64/orphanage/Orphanage.thy:2163]: This is probably mostly just copied, but I'll have a go at crunching it.

**Comment by lsf37** [comment]
All review feedback addressed, commit messages fixed up, and branch is rebased. Will merge when the tests are done.

**Pull request merged**
